### ISSUE TITLE
Flutter web cannot render high resolution images which are resolution aware

### ISSUE THREAD

**Issue by kenNg1** [flutter/flutter#109109]
Flutter web cannot render high resolution images which are resolution aware. I am using Flutter to build web/mobile apps in one codebase for large clients at our consultancy so it's important that this issue gets fixed so that we can build a high-quality website.

Linked issues and other people facing issues:
https://github.com/flutter/flutter/issues/41328 - root cause identified
https://github.com/flutter/engine/pull/17209 - hasn't fixed the issue
https://stackoverflow.com/questions/67716900/flutter-web-pictures-get-pixelated - stack overflow others facing issue (viewed 700+ times)

Root cause:
`devicePixelRatio` in web is always 1 which means only the lowest resolution image will be rendered. 

My app and code:

<img width="1432" alt="image" src="https://user-images.githubusercontent.com/25297786/183286804-c22d8a58-f85a-4a4c-b713-e68474b95b3b.png">
Note: you need to right click the above image and open in new tab to see pixelation issue (as the image is very small on Github). If i hardcode the path to a 2x or 3x image, the image looks crisp so it's definitely a problem.

![image](https://user-images.githubusercontent.com/25297786/183286978-9d2fc458-53f6-4431-8e17-589ac40aea96.png)

```
  assets:
    - assets/images/
    - assets/images/2.0x/
    - assets/images/3.0x/
  ```

```
Widget build(BuildContext context) {
    final MediaQueryData queryData = MediaQuery.of(context);
    double devicePixelRatio = queryData.devicePixelRatio;
    print(devicePixelRatio);  // Console prints devicePixelRatio on Web as 1

    return ...
                    Image(
                      width: 412.w,
                      fit: BoxFit.fitWidth,
                      image: const AssetImage(
                        "assets/images/splash-doctors.png",
                      ),
                    ),
                ...
```

**Comment by OutdatedGuy** [comment]
Yo @kenNg1, I just have one doubt.

Can you check if the same happens on `Firefox` browser as well.

Build the demo app for Web > Host it somewhere like on Netlify > Try it on `Firefox`.

Please also share the hosted demo's link.

**Comment by kenNg1** [comment]
first up - firefox also has the same problem.
![image](https://user-images.githubusercontent.com/25297786/183363798-28e8e8a2-700f-4100-bf09-8dfb36058b61.png)

we will host onto netlify shortly - thanks

**Comment by exaby73** [comment]
Hi @..., please provide a [minimal, reproducible example][1] and the output of `flutter doctor -v`.

[1]: https://stackoverflow.com/help/minimal-reproducible-example

**Comment by kenNg1** [comment]
1x image
![splash-doctors](https://user-images.githubusercontent.com/25297786/183571347-e3534fe8-d353-41d3-82ec-9ad7fb649db0.png)

2x image
![splash-doctors](https://user-images.githubusercontent.com/25297786/183571435-3fb7cfdd-c9eb-4a89-9d20-1b0d47b5aef6.png)

3x image
![splash-doctors](https://user-images.githubusercontent.com/25297786/183571495-82811689-32ed-40ff-a667-dad2dc078650.png)

Setup images like this:
![image](https://user-images.githubusercontent.com/25297786/183571594-0f1662a4-5fed-4d13-aab0-8d2bbb2c406a.png)

main.dart:
```
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(home: MyHomePage());
  }
}

class MyHomePage extends StatelessWidget {
  const MyHomePage({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: const <Widget>[
            Image(
              width: 800,
              fit: BoxFit.cover,
              image: AssetImage("assets/images/splash-doctors.png"),
            )
          ],
        ),
      ),
    );
  }
}
```

**Comment by kenNg1** [comment]
```
[✓] Flutter (Channel stable, 3.0.5, on macOS 12.5 21G72 darwin-x64, locale en-HK)
    • Flutter version 3.0.5 at /Users/keng/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision f1875d570e (4 weeks ago), 2022-07-13 11:24:16 -0700
    • Engine revision e85ea0e79c
    • Dart version 2.17.6
    • DevTools version 2.12.2

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at /Users/keng/Library/Android/sdk
    • Platform android-33, build-tools 33.0.0
    • ANDROID_HOME = /Users/keng/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)

[✓] IntelliJ IDEA Ultimate Edition (version 2022.1.4)
    • IntelliJ at /Applications/IntelliJ IDEA.app
    • Flutter plugin version 69.0.4
    • Dart plugin version 221.6008.17

[✓] VS Code (version 1.70.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension can be installed from:
      🔨 https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-x64     • macOS 12.5 21G72 darwin-x64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 104.0.5112.79

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

**Comment by exaby73** [comment]
I can reproduce this issue on stable and master. The image indeed is pixelated. Not sure if it's coming clear in the screenshot I shared, but if you zoom in a bit, you can see the pixels.

<details>
<summary>Screenshot</summary>

<img width="600" src="https://user-images.githubusercontent.com/50555895/183613151-9f1c2aa1-3818-4405-9631-30615819d60a.png">

</details>

<details>
<summary><code>flutter doctor -v</code> (Stable)</summary>

```
[✓] Flutter (Channel stable, 3.0.5, on Microsoft Windows [Version 10.0.22000.832], locale en-IN)
    • Flutter version 3.0.5 at C:\Users\Nabeel Parkar\fvm\versions\stable
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision f1875d570e (4 weeks ago), 2022-07-13 11:24:16 -0700
    • Engine revision e85ea0e79c
    • Dart version 2.17.6
    • DevTools version 2.12.2

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at C:\android-sdk
    • Platform android-33, build-tools 33.0.0
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 11.0.12+7-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • CHROME_EXECUTABLE = C:\Program Files\BraveSoftware\Brave-Browser\Application\brave.exe

[✓] Visual Studio - develop for Windows (Visual Studio Community 2022 17.2.6)
    • Visual Studio at C:\Program Files\Microsoft Visual Studio\2022\Community
    • Visual Studio Community 2022 version 17.2.32630.192
    • Windows 10 SDK version 10.0.19041.0

[✓] Android Studio (version 2021.2)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+7-b1504.28-7817840)

[✓] VS Code (version 1.70.0)
    • VS Code at C:\Users\Nabeel Parkar\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.46.0

[✓] Connected device (3 available)
    • Windows (desktop) • windows • windows-x64    • Microsoft Windows [Version 10.0.22000.832]
    • Chrome (web)      • chrome  • web-javascript • unknown
    • Edge (web)        • edge    • web-javascript • Microsoft Edge 90.0.818.66

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

</details>

<details>
<summary><code>flutter doctor -v</code> (Master)</summary>

```
[✓] Flutter (Channel master, 3.1.0-0.0.pre.2166, on Microsoft Windows [Version 10.0.22000.832], locale en-IN)
    • Flutter version 3.1.0-0.0.pre.2166 on channel master at C:\Users\Nabeel Parkar\fvm\versions\master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 87f1684476 (3 hours ago), 2022-08-08 23:00:11 -0700
    • Engine revision b6dd604d90
    • Dart version 2.19.0 (build 2.19.0-76.0.dev)
    • DevTools version 2.16.0

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at C:\android-sdk
    • Platform android-33, build-tools 33.0.0
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 11.0.12+7-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • CHROME_EXECUTABLE = C:\Program Files\BraveSoftware\Brave-Browser\Application\brave.exe

[✓] Visual Studio - develop for Windows (Visual Studio Community 2022 17.2.6)
    • Visual Studio at C:\Program Files\Microsoft Visual Studio\2022\Community
    • Visual Studio Community 2022 version 17.2.32630.192
    • Windows 10 SDK version 10.0.19041.0

[✓] Android Studio (version 2021.2)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+7-b1504.28-7817840)

[✓] VS Code (version 1.70.0)
    • VS Code at C:\Users\Nabeel Parkar\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.46.0

[✓] Connected device (4 available)
    • sdk gphone64 x86 64 (mobile) • emulator-5554 • android-x64    • Android 13 (API 33) (emulator)
    • Windows (desktop)            • windows       • windows-x64    • Microsoft Windows [Version 10.0.22000.832]
    • Chrome (web)                 • chrome        • web-javascript • unknown
    • Edge (web)                   • edge          • web-javascript • Microsoft Edge 90.0.818.66

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

</details>

**Comment by yjbanov** [comment]
Redirecting to the framework team for now. The logic for selecting image variants lives in the framework.

**Comment by goderbauer** [comment]
In the framework, we mainly use the devicePixelRatio to select the right asset variant. Looks like others on this issue have already narrowed down that the DPR available in the framework is incorrect. AFAIK, we pass the DPR through right from the engine.

@yjbanov Can you confirm that the web engine actually provides the framework with the correct DPR value? If yes, then yeah, we somewhere mess up asset selection based on that correct DPR value. However, if the DPR value is incorrectly reported by the engine, there's not much we can do on the framework side to magically select the right asset based on an incorrect DPR.

**Comment by OutdatedGuy** [comment]
I followed the setup using images with 1x, 2x and 3x density, sample code, and folder structure.
In my case, the app is using 2x image, which is expected for my screen. And changes the images when zoomed in and out.

<details>
<summary>Screenshots for <code>html</code> renderer</summary>

### For 80% (uses 1x)
![image](https://user-images.githubusercontent.com/74326345/184233551-2665f128-4222-4a9e-a0f6-203bb55cc53d.png)

### For 90% or 100% (uses 2x)
![image](https://user-images.githubusercontent.com/74326345/184233609-24b049d4-56dd-4a44-8849-5bb798310c93.png)

### For 200% (uses 2x)
![image](https://user-images.githubusercontent.com/74326345/184233640-5fd7534d-59a7-499c-804b-282758ce4482.png)

### For 240% or 250% (uses 3x)
![image](https://user-images.githubusercontent.com/74326345/184233683-a5acbf0b-4580-407b-b2cc-844c58c0cbf5.png)

</details>

<details>
<summary>Screenshots for <code>canvaskit</code> renderer</summary>

### For 80% (uses 1x)
![image](https://user-images.githubusercontent.com/74326345/184234435-9fe84057-3087-4bce-aa7a-822b9e296ed4.png)

### For 90% or 100% (uses 2x)
 ![image](https://user-images.githubusercontent.com/74326345/184234470-a93b42db-961e-4123-aadd-a442e332a64f.png)

### For 200% (uses 2x)
![image](https://user-images.githubusercontent.com/74326345/184234585-029f1ded-706d-48d1-81bf-cbfddd0f8393.png)

### For 240% or 250% (uses 3x)
![image](https://user-images.githubusercontent.com/74326345/184234643-657cea3c-6cc0-4f9e-b96e-ffdbf5bbe723.png)

</details>

It is working as expected for me or am I missing something?

**Comment by kenNg1** [comment]
@OutdatedGuy thanks for trying this - the issue still persists on bigger monitors - I have two 27" Samsung monitors (slightly different models - one is Essential Monitor T35F).

However, I was able to reproduce the same results as you using my 13 inch mac screen. 

Evidence of issue on large monitors (e.g. my Samsung monitors):

<details>
 <summary>Screenshots of `canvaskit` renderer</summary>
 ![Screenshot 2022-08-13 at 10 13 52 AM (3)](https://user-images.githubusercontent.com/25297786/184464758-164aeb51-a338-4305-b0ab-dfa371b3cd59.png)
</details>

<details>
 <summary>Flutter inspector showing 1x image used for above image</summary>
![image](https://user-images.githubusercontent.com/25297786/184464805-b59b75e2-f556-4ab7-b0cc-eebccc6a9c56.png)
</details>

**Comment by goderbauer** [comment]
@kenNg1 What are the native device pixel ratios on these screens you're using? Executing `window.devicePixelRatio` in your browser's console while the browser window is displayed on the screen in question should give the answer.

**Comment by yjbanov** [comment]
Something to be aware of: `window.devicePixelRatio` = `native device pixel ratio` x `page zoom level`. For example, on a typical mac laptop screen the native device pixel ratio is 2.0. If you zoom the page to 110%, the page's effective device pixel ratio will be 2.2, and that's the value that will be reported to the framework.

In [this screenshot](https://user-images.githubusercontent.com/25297786/184464758-164aeb51-a338-4305-b0ab-dfa371b3cd59.png) the macOS text rendering looks blurry. Perhaps the system is not configured to match the display's native resolution?

**Comment by kenNg1** [comment]
@yjbanov thanks for the note.

It is still affecting the `stable` channel, however, I've tried the latest `beta` channel and it seems to be rendering 2x, 3x images as required (even when I zoom in)

[Probably related to this issue, but different symptom](https://github.com/flutter/flutter/issues/77605)

I will keep this ticket open until I have verified it works in stable channel.

**Comment by kenNg1** [comment]
Since this is fixed in beta, can the commit not be cherry picked and released as a hotfix? It seems like quite a critical bug for web because it affects the UI when you zoom in / out, it breaks adaptive images and it causes the screen to go white when moving the screen from a laptop monitor to a larger monitor.

I can confirm it works in beta so can a hotfix be explored?

**Comment by kenNg1** [comment]
Fixed with 3.30 - thank you!
